### ISSUE TITLE
Refreshing related records after updating

### ISSUE THREAD

**Issue by hdijkema** [espocrm/espocrm#1694]
I'm working on a quick prototype, extending the Formula functionality being able to process more hooks. See attached Formula.php hook. 

See added functionality here:

* [while functionality in formulas ](https://github.com/hdijkema/espocrm/blob/master/application/Espo/Core/Formula/Functions/WhileType.php)
* [formulas that can handle multiple hooks instead of only 'beforeSave'](https://github.com/hdijkema/espocrm/blob/master/application/Espo/Hooks/Common/Formula.php)

In my formula, I trigger the recalculation of a related record, using an update of the record:

```
begin:afterSave
$dt = datetime\now();
$s = datetime\format($dt, 'Europe/Amsterdam', 'MM/DD/YYYY hh:mm:ss');
$jr = jaar;
$rec_id = record\findOne('Nascholingstotaal', 
                             'createdAt', 'asc', 
                             'jaar=', $jr, 'contactId=', contactId);
record\update('Nascholingstotaal', $rec_id, 
              'description', string\concatenate("AFTER SAVE ", $s)
              );
end:afterSave

begin:afterRemove
$dt = datetime\now();
$s = datetime\format($dt, 'Europe/Amsterdam', 'MM/DD/YYYY hh:mm:ss');
$jr = jaar;
$rec_id = record\findOne('Nascholingstotaal', 
                             'createdAt', 'asc', 
                             'jaar=', $jr, 'contactId=', contactId);
record\update('Nascholingstotaal', $rec_id, 
              'description', string\concatenate("AFTER REMOVE ", $s)
              );
end:afterRemove
```

The related record 'Nascholingstotaal' is recalculated nicely. The following formula is being processed:

```
$ds = description;
$jr = jaar;
$contact_id = contactId;

$ids = record\findRelatedMany('Contact', contactId, 
                              'nascholingen', 1000, 
                              'createdAt', 'asc');
                              
$i = 0;
$n = array\length($ids);
$comma = "";

$totaal_dagen = 0;
while(
    $i < $n,
    $comma = ", ";
    $nascholing_id = array\at($ids, $i);
    $rec_jr = record\attribute('Nascholing', $nascholing_id, 'jaar');
    ifThen($rec_jr == $jr,
        $dagen = record\attribute('Nascholing', $nascholing_id, 'aantalDagen');
        $totaal_dagen = $totaal_dagen + $dagen;
    );
    $i = $i + 1;
);

dagen = $totaal_dagen;
```

However, the related entity is Nascholingstotaal is not refreshed on the screen.  See image.

![image](https://user-images.githubusercontent.com/745693/80916986-7523a780-8d5c-11ea-9950-7ed88e3d718a.png)

So, the blue record in 'Nascholing' has just been updated. All went well, all formulas have been calculated. Totals have been recalculated; however, it doesn't show. See the red rectangle.

When I manually refresh the screen, it does get the right result:

<img src="https://user-images.githubusercontent.com/745693/80917046-c7fd5f00-8d5c-11ea-9efb-a975aade4518.png" width="400">

**It is possible to trigger a refresh from the PHP backend code?**

**Comment by yurikuzn** [comment]
> It is possible to trigger a refresh from the PHP backend code?

No.

You can define a custom view for your relationship panel.  See example: https://github.com/espocrm/espocrm/blob/5.8.5/application/Espo/Modules/Crm/Resources/metadata/clientDefs/Account.json#L66

Create a js file for the view in client. In the ` setup` method:

```js
setup: function () {
    this.listenTo(this.model, 'sync', function () {
        this.collection.fetch();
    }, this);
},
```

**Comment by hdijkema** [comment]
Thank you for the pointers, AFAIK I did what you pointed out, but maybe I'm missing something. This is my json:

In custom/Espo/Custom/Resources/metadata/clientDefs/Contact.json

```json
{
    "kanbanViewMode": false,
    "color": "#a4c5e0",
    "iconClass": "fas fa-id-badge",
    "relationshipPanels": {
        "nascholingstotalen": {
            "select": false,
            "create": false,
            "view": "custom:views/record/panels/nascholingstotaal-panel"
        }
    }
}
```

In client/custom/src/views/record/panels/nascholingstotaal-panel.js:

```javascript
define(
    'custom:views/record/panels/nascholingstotaal-panel',
    [ 'views/record/panels/relationship' ],
    function (Dep) {
        return Dep.extend({
                  setup: function() {
                             alert('oke!');
                             this.listenTo(this.model, 'sync',
                                           function() {
                                               this.collection.fetch();
                                           },
                                           this
                                          );
                          },
                  }
               );
     }
);
```

I figure, this definition extends the relationship panel? Is can see `setup` is called, because it alerts me with `'oke!'`.  

However, when I refresh the Contact screen, it shows an empty panel now:

![image](https://user-images.githubusercontent.com/745693/80945323-dcd50380-8deb-11ea-82b1-c08511a9f113.png)

Am I forgetting something? 

Without the custom view, it shows the relationship table.

**Comment by hdijkema** [comment]
Oh, I see: Need to call `Dep.prototype.setup.call(this)`.

**Comment by hdijkema** [comment]
Under what circumstances is 'sync' called?

**Comment by yurikuzn** [comment]
When model is fetched or saved.

**Comment by hdijkema** [comment]
> When model is fetched or saved.

I see that 'sync' is called for 'nascholing' when I save a record. From this point on, I would want to trigger a 'sync' on 'nascholingstotalen'. However, this is an other portal on the page. 

![image](https://user-images.githubusercontent.com/745693/80953127-3a248100-8dfb-11ea-8768-31ac279f44dd.png)

**Comment by yurikuzn** [comment]
Model is passed from the parent view. Both panels share the same parent views. You can listen to 'sync' event in both panels. It's not good to trigger 'sync' since it's an event with its own purpose. You can trigger any other event like `this.trigger('my-event')`.

**Comment by hdijkema** [comment]
This sync method is only called when I modify the 'Contact', not when I modify, add or remove a 'Nascholing'. 

The setup methods of the bottom panels are called with the model for 'Contact'. So, when I change something in the bottom panel, I see that the sync method is called in the model for 'nascholingen', but that is not the model that I'm listening to. 

I added some alerts to see what's going on. This happens when I edit a 'nascholing'.

<img src="https://user-images.githubusercontent.com/745693/80959131-74475000-8e06-11ea-9b22-a8c829219323.png" width="400">

<img src="https://user-images.githubusercontent.com/745693/80959168-875a2000-8e06-11ea-9050-8f4cfcd31293.png" width="400">

No 'sync' listener is called.

When the page sets up, I see following:

<img src="https://user-images.githubusercontent.com/745693/80959286-c6887100-8e06-11ea-838a-f02f20750834.png" width="400">

I see my panel setup called with a Contact. And it registers a 'sync' event handler on this Contact. 
However, Contact will never be changed when I change a Nascholing, hence the 'sync' event handler of the panel doesn't get called. 

![image](https://user-images.githubusercontent.com/745693/80959772-c76dd280-8e07-11ea-9b3d-c2a5cd9f1e0f.png)

**Comment by yurikuzn** [comment]
I looked into the code and found two events called on model:
* after:relate
* after:unrelate

Try to listen on both events in your second relationship panel.

**Comment by yurikuzn** [comment]
But beware of possible looping. These events will be triggered also when actions are done for the same panel.

**Comment by hdijkema** [comment]
I see, thank you. It almost works. It is called on any relate or unrelate to the model (the Contact), so that works. However, when I update a relation (e.g. by selecting a new record), no relate or unrelate or e.g. 'relation-modified' is triggered.

**Comment by yurikuzn** [comment]
Do you mean when you edit the related record? There's no any event. Only 'sync' is called for the collection when it's re-fetched after update.

**Comment by hdijkema** [comment]
Yes, I'm editing the record. What do you mean by 'sync' called for the collection? When I add a trigger on the parent model in the 'list.js' view, just after line 1731:

```
                        var parent_model = model.collection.parentModel;
                        parent_model.trigger('after:related-saved', m);
```

I can listen to it in my view:

```
                                this.listenTo(this.model, 'after:related-saved',
                                              function() {
                                                  this.model.trigger('totaal-refresh');
                                              }, this
                                            );
```

But that is probably not the best way to go...Who says this list represents a set of related records? 

On setup, the collection is undefined, so I do not seem to be able to listen to the sync of the collection?

**Comment by yurikuzn** [comment]
You can do it in setFilter method, the only one available.  See: https://github.com/espocrm/espocrm/blob/5.8.5/client/src/views/record/panels/relationship.js#L181

**Comment by hdijkema** [comment]
Hmm. Collection is indeed there, but I'm not sure that I'm doing the right thing here. 

I'm calling the parent setFilter and then I add a sync listener to the collection.  It does get called when the page (re)builds, but not when the collection is re-fetched after update.

```javascript
                  setFilter: function (filter) {
                        Dep.prototype.setFilter.call(this, filter);
                        console.log('setFilter');
                        console.log(this);
                        console.log(this.model);
                        console.log(this.collection);
                        console.log('endSetFilter');
                        this.listenTo(this.collection, 'sync', function() { alert('Hi!'); }, this);
                        },
                  }
```

**Comment by yurikuzn** [comment]
I'm not sure. Here where fetch is suppose to be called after you edited: https://github.com/espocrm/espocrm/blob/5.8.5/client/src/views/record/panels/relationship.js#L475

**Comment by hdijkema** [comment]
I see. Yes I tried to override this method already, but as a matter of fact, `actionEditRelated` is never called. Instead, the method `actionQuickEdit` of list.js gets called. 

This seems to result in some strange behaviour. E.g. when I edit 'member 1' and change a relation as follows:

From member 1:

![image](https://user-images.githubusercontent.com/745693/80976184-6653f800-8e23-11ea-83db-2f2995727539.png)

to member 2:

![image](https://user-images.githubusercontent.com/745693/80976290-8c799800-8e23-11ea-8ea8-683c5d9e8269.png)

The list of related records still seems to contain the same record, although it is not related to member 1 anymore. A refresh will solve the issue (as does the method you refer to).

**Comment by hdijkema** [comment]
I think this it is tricky to get this behaviour right at the client side, while actually the updates are being done with a formula, so I implemented a simple signalling mechanism.

https://github.com/hdijkema/espocrm/commit/3708c3018688be90ec56c4c2251896b768f753d3

I'll probably make an extension package for it.  It needs some extra work. And it would be great if the javascript WeakRef would be available.

So now, this is the 'nascholingstotaal-panel.js', working with signals:

```javascript
/* vi: set sw=4 ts=4:  */
define(
    'custom:views/record/panels/nascholingstotaal-panel',
    [ 'views/record/panels/relationship', 'signals' ],
    function (Dep, Signals) {
        return Dep.extend({
                    setup: function() {
                        Dep.prototype.setup.call(this);
                        var signals = Signals;
                        var me = this;
                        signals.registerSignal(
                                            this.model.id,
                                            function() { me.collection.fetch(); },
                                            2);
                    },
                  }
               );
     }
);
```

And this is the formula, working with the signal (after-remove excerpt):

```
begin:afterRemove
$dt = datetime\now();
$s = datetime\format($dt, 'Europe/Amsterdam', 'MM/DD/YYYY hh:mm:ss');
$jr = jaar;
$rec_id = record\findOne('Nascholingstotaal', 
                             'createdAt', 'asc', 
                             'jaar=', $jr, 'contactId=', contactId);
record\update('Nascholingstotaal', $rec_id, 
              'description', string\concatenate("AFTER REMOVE ", $s)
              );
signal(contactId);
end:afterRemove
```

signal(contactId) triggers the client panel to refetch it's contents.

**Comment by yurikuzn** [comment]
Such signals will require websocket usage. It's already done for the stream panel.

* Subscribe in frontend: https://github.com/espocrm/espocrm/blob/5.8.5/client/src/views/stream/panel.js#L209
* Submit in backend: https://github.com/espocrm/espocrm/blob/5.8.5/application/Espo/Hooks/Note/WebSocketSubmit.php#L56

You need to have websocket configured on your server: https://docs.espocrm.com/administration/websocket/

**Comment by hdijkema** [comment]
Thanks! I'll implement the extension this way.
